File: baekjoon_contest/096. (231118) 제1회 스타보우컵/2_Orange_FP_boj_30677.cpp
// Solve 2023-11-18

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int main() {
    FASTIO;

    int n, k, c, r;
    cin >> n >> k >> c >> r;

    vector<int> base(k + 1, 0), s(k + 1, 0), p(k + 1, 0);

    for (int i = 1; i <= k; i++) {
        cin >> base[i];
    }

    for (int i = 1; i <= k; i++) {
        cin >> s[i];
    }

    for (int i = 1; i <= k; i++) {
        cin >> p[i];
    }

    ll stardust = 0;
    int combo = 0, fatigue = 0;
    vector<int> skill(k + 1, 0);

    for (int ni = 1; ni <= n; ni++) {
        int l;
        cin >> l;

        if (stardust == -1) continue;

        if (l == 0) {
            combo = 0;
            fatigue -= r;

            if (fatigue < 0) {
                fatigue = 0;
            }
        }
        else {
            stardust += (ll)(100 + combo * c) * (100 + skill[l] * s[l]) * base[l] / 10000;
            combo++;
            skill[l]++;
            fatigue += p[l];

            if (fatigue > 100) {
                stardust = -1;
            }
        }
    }

    cout << stardust << '\n';

    return 0;
}

File: baekjoon_contest/2023/096_231118_제1회 스타보우컵/2_Orange_FP_boj_30677.py
# Solve 2023-11-18

import sys

input = lambda : sys.stdin.readline().rstrip()

n, k, c, r = map(int, input().split())
base = list(map(int, input().split()))
s = list(map(int, input().split()))
p = list(map(int, input().split()))

stardust = 0
combo = 0
fatigue = 0
skill = [0 for _ in range(k)]

for _ in range(n):
    l = int(input()) - 1

    if stardust == -1:
        continue

    if l == -1:
        combo = 0
        fatigue -= r

        if fatigue < 0:
            fatigue = 0
    else:
        stardust += (100 + combo * c) * (100 + skill[l] * s[l]) * base[l] // 10000
        combo += 1
        skill[l] += 1
        fatigue += p[l]

        if fatigue > 100:
            stardust = -1

print(stardust)
